File: src/main/java/net/sf/rails/common/parser/GameOptionsParser.java
package net.sf.rails.common.parser;

import java.io.File;
import java.util.List;
import java.util.Map;

import net.sf.rails.common.GameOption;
import net.sf.rails.common.GameOptionsSet;

import org.w3c.dom.Document;
import org.w3c.dom.Element;

import com.google.common.base.Splitter;


public class GameOptionsParser {

    private final static String FILENAME = "GameOptions.xml";
    private final XMLParser parser = new XMLParser();

    public GameOptionsParser() {}

    public GameOptionsSet.Builder processOptions(String directory) throws ConfigurationException {

        GameOptionsSet.Builder options = GameOptionsSet.builder();
        
        Document doc = parser.getDocument(FILENAME, directory);
        Element root = parser.getTopElement(doc);

        List<Element> elements = parser.getElementList(XMLTags.OPTION_TAG, root.getChildNodes());
       
        // use ordering provided in the xml-file
        int ordering = 0;
        for (Element element:elements) {
            Map<String, String> optionMap = parser.getAllAttributes(element);

            GameOption.Builder option;
            if (optionMap.containsKey(XMLTags.NAME_ATTR)) {
                option = GameOption.builder(optionMap.get(XMLTags.NAME_ATTR));
            } else {
                option = null;
            }

            if (option != null) {
                option.setOrdering(ordering++);
                
                if (optionMap.containsKey(XMLTags.TYPE_ATTR)) {
                    option.setType(optionMap.get(XMLTags.TYPE_ATTR));
                }

                if (optionMap.containsKey(XMLTags.DEFAULT_ATTR)) {
                    option.setDefaultValue(optionMap.get(XMLTags.DEFAULT_ATTR));
                }

                if (optionMap.containsKey(XMLTags.PARM_ATTR)) {
                    String parameters = optionMap.get(XMLTags.PARM_ATTR);
                    option.setParameters(Splitter.on(XMLTags.VALUES_DELIM).split(parameters));
                }
                
                if (optionMap.containsKey(XMLTags.VALUES_ATTR)) {
                    String values = optionMap.get(XMLTags.VALUES_ATTR);
                    option.setAllowedValues(Splitter.on(XMLTags.VALUES_DELIM).split(values));
                }
                options.add(option.build());
            }
        }
        return options;
    }

    public static GameOptionsSet.Builder load(String gameName) throws ConfigurationException {
        GameOptionsParser gop = new GameOptionsParser();
        String directory =  GameInfoParser.DIRECTORY + File.separator + gameName;
        return gop.processOptions(directory);
    }

}

File: dotnet/RailsLib.Net/Net/Common/Parser/GameOptionsParser.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace GameLib.Net.Common.Parser
{
    public class GameOptionsParser
    {
        private const string FILENAME = "GameOptions.xml";
        private XmlParser parser = new XmlParser();

        public GameOptionsParser() { }

        public GameOptionsSet.Builder ProcessOptions(string directory)
        {
            GameOptionsSet.Builder options = GameOptionsSet.GetBuilder();

            XmlDocument doc = parser.GetDocument(FILENAME, directory);
            XmlElement root = parser.GetTopElement(doc);

            List<XmlElement> elements = parser.GetElementList(XmlTags.OPTION_TAG, root.ChildNodes);

            // use ordering provided in the xml-file
            int ordering = 0;
            foreach (XmlElement element in elements)
            {
                Dictionary<string, string> optionMap = parser.GetAllAttributes(element);

                GameOption.Builder option;
                if (optionMap.ContainsKey(XmlTags.NAME_ATTR))
                {
                    option = GameOption.CreateBuilder(optionMap[XmlTags.NAME_ATTR]);
                }
                else
                {
                    option = null;
                }

                if (option != null)
                {
                    option.SetOrdering(ordering++);

                    if (optionMap.ContainsKey(XmlTags.TYPE_ATTR))
                    {
                        option.SetType(optionMap[XmlTags.TYPE_ATTR]);
                    }

                    if (optionMap.ContainsKey(XmlTags.DEFAULT_ATTR))
                    {
                        option.SetDefaultValue(optionMap[XmlTags.DEFAULT_ATTR]);
                    }

                    if (optionMap.ContainsKey(XmlTags.PARM_ATTR))
                    {
                        string parameters = optionMap[XmlTags.PARM_ATTR];
                        option.SetParameters(parameters.Split(XmlTags.VALUES_DELIM));
                    }

                    if (optionMap.ContainsKey(XmlTags.VALUES_ATTR))
                    {
                        string values = optionMap[XmlTags.VALUES_ATTR];
                        option.SetAllowedValues(values.Split(XmlTags.VALUES_DELIM));
                    }
                    options.Add(option.Build());
                }
            }
            return options;
        }

        public static GameOptionsSet.Builder Load(string gameName)
        {
            GameOptionsParser gop = new GameOptionsParser();
            // use the Separator provided by Resource-Loader!
            string directory = GameInfoParser.DIRECTORY + ResourceLoader.SEPARATOR + gameName;
            return gop.ProcessOptions(directory);
        }
    }
}
